accepted/future-releases/class-modifiers/feature-specification.md

```diff
@@ -836,38 +785,82 @@ the latter, even if the class is being used as a mixin in a post-feature
 library where it does happen to be possible to distinguish those two
 intents.*
 
+### Enum classes
+
+The class introduced by an `enum` declaration is considered `final` for any
+purpose where a class modifier is required.
+
+The behavior of `enum` declarations is unchanged. Since an `enum` class cannot
+have any subclasses, the implicit `final` modifier would not prevent any
+otherwise allowed operation.
+The implicit `final` also does not imply fewer restrictions than `enum`
+declarations would otherwise have. The modifier is applied only to automatically
+satisfy any requirements introduced by super-interfaces.
+
 ### Anonymous mixin applications
```

Comment:
This section is super confusing.  I *believe* that this has no relevance to end users (or even implementers per se, at least for the static semantics) because it purely is about how the de-sugaring of mixins works, correct?  That is, there is no information an end user needs to know about here, and no errors to check for here.  Assuming that is correct, I think it would really help readers of this spec (i.e. our implementation teams) to have that context.  I'd say something at the start of this section like: "The specification of the semantics of mixins is given in terms of a rewrite in terms of a sequence of anonymous mixin classes.  The following section describes how to do that rewrite in a way that satisfies the restrictions described in this feature spec.  This section is purely informational, and specifies no new user facing errors or warnings".

Reply:
I think we can remove the section, if we are *very careful* to not say something like "Every subclass of a `base` class is marked `base` or ...", because an anonymous mixin application is not a declaration and is therefore not marked anything.
And for `sealed`, it's even more important to get things just right.

What I really want is to to introduce a semantic abstraction layer, so our semantics don't have to talk directly about syntax.
I want to say something like: 

> An interface is *sealed* if it's the interface of a declaration marked `sealed` or if it's an anonymous mixin application whose superclass or mixin's interface is *sealed*.

and then, in the definition of exhaustiveness, never ask about the `sealed` modifier, only the *sealed* semantic property.

Also:

> An interface _prevents implementation_ if it is the interface of a declaration marked `base` or `final`,
> or if it has any superinterface which _prevents implementation_.
> An interface _prevents local implementation in library L_ if it has a superinterface which prevents 
> implementation, and that superinterface's declaration is in a library other than *L*.
> A class or mixin _prohibits inheritance_ if its declaration is marked `final` or `base`.

Then the errors become:
* It's a compile-time error if a class, mixin, or mixinclass declaration has 
   an interface which prevents implementation, and the declaration is not marked `base`, `final` or `sealed`.
* It's a compile-time error to extend a class or mix-in a mixin which prohibits inheritance,
   or whose interface is sealed, in another library than where the class or mixin is declared.
* It's a compile-time error to implement an interface in library *L* if
    * the interface prevents implementation in *L*, or if
    * the interface is sealed and not declared in *L*.
* It's a compile-time error for a mixin to have an `on` type interface which is sealed and
   declared in another library that the mixin.

This all works without having to consider whether an anonymous mixin application has a modifier 
or not. It has a declaration, but its interface is not "the interface" of that declaration.
All these little distinctions between declaration, declaration of, interface of a (class or declaration),
I want to get them *right*, because otherwise the specification risks being subtly ambiguous.


This section is ... doing something intermediate, which is probably why it's confusing.
It introduces the semantic properties, then it uses that to add synthetic modifiers to things that are not really declarations, because the rest of the specification *is* talking about things in terms of, say, classes having modifiers (they don't, their declaration might, but anonymous mixin applications don't have the modifiers of their declaration), and I want to be compatible with the rest.

(We cannot just say that the anonymous mixin application have the modifiers of the declaration it's in, because
then the following would not make `S&M` be `sealed`, and then it's not obvious that `case A()` would help exhausting `S`.
```dart
sealed class S {}
class A extends S with M {}
```
so ... need to be precise!)

Reply:
I'm more lost than I started.  I honestly have no idea what you're both trying to accomplish with this section.  I completely believe that there is a reason for it, but ... I don't know what it is, and so I'd suggest that probably no other readers do either.  I *think* (but am no longer sure?) that this section is intended to describe (in an indirect way) what modifiers you would put on the implicitly generated classes that you would generate if you were to follow the desugaring semantics of mixins, instead of implementing a direct (static or runtime) semantics.  I suspect however that neither of our implementations use the desugaring for the static semantics, and I believe that backends can choose whether or not to use the desugaring for the dynamic semantics (and in any case, this is irrelevant for the dynamic semantics).  So I believe that this section is entirely irrelevant except to make the desugaring semantics used elsewhere in the spec correct.  Is that belief true?  If not, it's *really important* that we clarify how it is not.  Because if this is intended to *add* some static errors, I don't know what they are (and hence I suspect none of the folks responsible for implementing this know either).

Reply:
If my hypothesis above is correct, that this is just about the desugaring, why not just explain it that way directly?  That is, give the desugaring, and say something like:

```
In either case, let $K$ be a class declaration with
the same constructors, superclass, interfaces and instance members as
the defined class.  Then $K$ has the class modifier <foo> if <bar>
```

??

Reply:
My radical opinion is that the entire notion of "anonymous mixin application classes" is an unnecessary attempt to cram a trait-like behavior square peg into a Smalltalk everything-is-a-class single inheritance round hole and causes more trouble than its worth in the spec.

I'm fairly certain there is a simpler way to incorporate mixins into the language without resorting to conjuring up a synthetic invisible class every time you use one. (And, also, mixin application classes are a useless feature as far as users are concerned.)

Reply:
I don't see any particular reason to connect the rules about implicit class modifiers of mixin applications with the desugaring semantics (that's the kind of semantics we actually have in the language specification; surely we could have others, but I don't think that would make any difference here).

The way I see it could be illustrated by the following two graphs. First, the underlying code:

```dart
interface class A {}
interface class B {}
mixin M implements A {}
sealed class C extends B with M {}
class D extends C {}
```

Next comes the superinterface graph, showing the explicit annotations (the dotted arrow is `implements`, the black arrow is `extends`):

```mermaid
graph BT;
  D --> C[sealed C]
  C --> BwM[B with M]
  BwM --> B[interface B]
  BwM -.-> M
  M -.-> A[interface A]
```

Actually, we had `A`, `B`, `C`, and `M` already, and I'm currently writing `D`. So I'm wondering, can I do it as shown above (`class D extends C {}`), or would that be an implicit reopening? In order to be able to reason about the code that I'm writing (as opposed to "try all combinations of keywords until there is no error"), I'd like to explain what's going on.

The answer is that it _is_ an implicit reopening, so we'll get the `implicit_reopen` lint.

Let's just say that we don't want that. Would it help to remove `interface` from `A`? How about `B`?

Here is the superinterface graph again, with implicitly induced modifiers:

```mermaid
graph BT;
  D --> C[sealed (interface) C]
  C --> BwM[(interface) B with M]
  BwM --> B[interface B]
  BwM -.-> M
  M -.-> A[interface A]
```

I'd argue that we can now use some simple rules to reason about the situation: Implicitly induced modifiers are always sufficient to avoid reopening, so we don't need to worry about them. However, `interface` should always be propagated across black arrows (`extends`).

Inspecting the graph, we can see that `interface` is not propagated from `C` to `D`, and that's the reason why we get the lint. I'd argue that the same thing is not entirely as obvious if we look at the previous graph.

Returning to the question: Would it help if we remove `interface` from `A`? In this case we can immediately see that no `interface` modifier is present on `M` (we don't need to propagate `interface` across a dotted line). This implies that we can't solve the problem by removing `interface` from `A`, the graph stays the same everywhere else.

However, if we remove `interface` from `B` then we get the following:

```mermaid
graph BT;
  D --> C[sealed C]
  C --> BwM[B with M]
  BwM --> B
  BwM -.-> M
  M -.-> A[interface A]
```

We can see that there is no reopening any more, because `C` isn't implicitly `interface` any more&mdash;so there's no need to propagate it, and hence `D` can be a plain class.

I find it easier to understand how the classes and their modifiers interact if we allow ourselves to show all of them (explicit + implicit), such that we can answer every question about possible usages of the given declaration, and questions about reopening, using local information only.

If we insist that only the information in the first graph can be shown to the world, and developers out there can't even learn the rules about how the implicit modifiers are computed for mixin applications, then I honestly think it gets harder to answer questions like the ones I mentioned above.

Reply:
> My radical opinion is that the entire notion of "anonymous mixin application classes" is an unnecessary attempt to cram a trait-like behavior square peg into a Smalltalk everything-is-a-class single inheritance round hole and causes more trouble than its worth in the spec.

That may be, but it's what we have today, and changing it is potentially breaking (because our naive LUB depends on type hierarchy depth). 
At least our current model is *simple*.

Now, about how to phrase things.

We should decide on the goal for this specification:
* Is it about being precise and formal, or
* Is it intended for end-users to read?

I'm going for formal. It should still be readable and understandable to implementors (and it may be failing that), but end-users should read a carefully hand-crafted description of the feature instead.

That also means that the phrasing should be aimed at implementors, so saying up front how every section applies to implementations is a good idea. That's missing here, where the specification is more like a free-floating abstract specification. It may be mathematically sound, but not very actionable. 
Note taken!



I want to avoid desugaring semantics. Desugaring semantics always leaves *me* scratching my head about what they mean, because it's unclear what status the desugared syntax has, and it seems that's often unclear to the author who wrote it too. (Even if that was me.) 
Is the original code removed from the program, and the desugaring put in its place? Does this happen before, during or after type inference? Does the new code have properly assigned context types and static types? (If so, which?) Very often running type inference on the desugared code again would give a different result.
What if the desugared code has an error, does that mean the original had one too?
Or can the desugared code ignore some requirements, _because_ it's actually the desugaring of the feature introducing that requirement? (If we desugar `enum` classes into normal classes with normal `const` constructor invocations, those classes _do_ extend `Enum` and are instantiable, which is otherwise an error.)

Lifting things to a semantic level before (or while) doing "rewrites" allows us to ignore syntactic restrictions.

Reply:
But also, yes, I believe the section is unnecessary if we are careful about how we specify things.
We cannot just say "if the superclass has the modifier ..." when the superclass might not have any modifiers, because it's an anonymous mixin application class.

I'm willing to try!

Reply:
And done. Things are now defined entirely in terms of super-declarations, not classes or interfaces (semantic entities, different from the syntactic declarations introducing them). Because anonymous mixin applications are just part of another declaration, they don't need further rules.